.travis.yml

```diff
@@ -1,195 +1,211 @@
-     cabal install --only-dependencies ${TEST} base-compat/ base-compat-batteries/ check/ --dry -v ${CABALFLAGS} > installplan.txt;
-   else
-     cabal install --only-dependencies base-compat/ --dry -v ${CABALFLAGS} > installplan.txt;
-   fi
- - sed -i -e '1,/^Resolving /d' installplan.txt; cat installplan.txt
-
-# check whether current requested install-plan matches cached package-db snapshot
- - if diff -u installplan.txt $HOME/.cabsnap/installplan.txt;
-   then
-     echo "cabal build-cache HIT";
-     rm -rfv .ghc;
-     cp -a $HOME/.cabsnap/ghc $HOME/.ghc;
-     cp -a $HOME/.cabsnap/lib $HOME/.cabsnap/share $HOME/.cabsnap/bin $HOME/.cabal/;
-   else
-     echo "cabal build-cache MISS";
-     rm -rf $HOME/.cabsnap;
-     mkdir -p $HOME/.ghc $HOME/.cabal/lib $HOME/.cabal/share $HOME/.cabal/bin;
-     if [ "x$TEST" = "x--enable-tests" ];
-     then
-       cabal install -j --only-dependencies ${TEST} ${CABALFLAGS} base-compat/ base-compat-batteries/ check/;
-     else
-       cabal install -j --only-dependencies base-compat/ ${CABALFLAGS};
-     fi
-   fi
-
-# snapshot package-db on cache miss
- - if [ ! -d $HOME/.cabsnap ];
-   then
-      echo "snapshotting package-db to build-cache";
-      mkdir $HOME/.cabsnap;
-      cp -a $HOME/.ghc $HOME/.cabsnap/ghc;
-      cp -a $HOME/.cabal/lib $HOME/.cabal/share $HOME/.cabal/bin installplan.txt $HOME/.cabsnap/;
-   fi
+  - cabal --version
+  - echo "$(${HC} --version) [$(${HC} --print-project-git-commit-id 2> /dev/null || echo '?')]"
+  - BENCH=${BENCH---enable-benchmarks}
+  - TEST=${TEST---enable-tests}
+  - HADDOCK=${HADDOCK-true}
+  - UNCONSTRAINED=${UNCONSTRAINED-true}
+  - NOINSTALLEDCONSTRAINTS=${NOINSTALLEDCONSTRAINTS-false}
+  - GHCHEAD=${GHCHEAD-false}
+  - travis_retry cabal update -v
+  - "sed -i.bak 's/^jobs:/-- jobs:/' ${HOME}/.cabal/config"
+  - rm -fv cabal.project cabal.project.local
+  - grep -Ev -- '^\s*--' ${HOME}/.cabal/config | grep -Ev '^\s*$'
+  - >
+    if [ "x$BATTERIES" = "xNO" ]; then
+      printf 'packages: ./base-compat\n' > cabal.project;
+    else
+      printf 'packages: ./base-compat ./base-compat-batteries ./check\n' > cabal.project;
+    fi
+  - touch cabal.project.local
+  - "if ! $NOINSTALLEDCONSTRAINTS; then for pkg in $($HCPKG list --simple-output); do echo $pkg  | grep -vw -- base-compat | grep -vw -- base-compat-batteries | grep -vw -- type-check | sed 's/^/constraints: /' | sed 's/-[^-]*$/ installed/' >> cabal.project.local; done; fi"
+  - cat cabal.project || true
+  - cat cabal.project.local || true
+  - if [ -f "./base-compat/configure.ac" ]; then
+      (cd "./base-compat" && autoreconf -i);
+    fi
+  - if [ -f "./base-compat-batteries/configure.ac" ]; then
+      (cd "./base-compat-batteries" && autoreconf -i);
+    fi
+  - if [ -f "./check/configure.ac" ]; then
+      (cd "./check" && autoreconf -i);
+    fi
+  - rm -f cabal.project.freeze
+  - travis_wait cabal new-build -w ${HC} ${TEST} ${BENCH} --project-file="cabal.project" --dep -j2 all
```

Comment:
compiling `Cabal` (for custom setups) on GHC-7.4.1

Reply:
Ah, I see. It's somewhat of a shame that `new-build` must rebuild `Cabal`, since it's an incredibly monolithic package. (I've also had AppVeyor builds time out due to `Cabal`.) I still don't know of a better solution than this one, though.

Reply:
good thing: it's cached.

I don't know the reason precisely, but new-build requires `Cabal >= 1.20` for custom setups, so resulting `./Setup` knows needed tricks (i.e. I don't know what exactly). This is a bit unfortunate (GHC 7.8.4 bundles Cabal-1.18 e.g.).

I'm (really slowly, not even with constant pace) working towards a solution, that we wouldn't need `build-type: Custom` for `doctest`s. The status quo is just acceptable enough :/